changeset/commit.go

```diff
@@ -17,113 +17,66 @@ limitations under the License.
 package changeset
 
 import (
-	"bufio"
-	"fmt"
-	"io/ioutil"
-	"os"
-	"path/filepath"
+	"errors"
 	"regexp"
-	"strings"
+	"runtime/debug"
+	"strconv"
+	"sync"
 )
 
-const (
-	commitIDFile      = "HEAD"
-	koDataPathEnvName = "KO_DATA_PATH"
-	// packedRefsFile is a file containing a list of refs, used to compact the
-	// list of refs instead of storing them on the filesystem directly.
-	// See https://git-scm.com/docs/git-pack-refs
-	packedRefsFile = "packed-refs"
-)
+const Unknown = "unknown"
+
+var (
+	shaRegexp = regexp.MustCompile(`^[a-f0-9]{40,64}$`)
+	rev       string
+	err       error
+	once      sync.Once
 
-var commitIDRE = regexp.MustCompile(`^[a-f0-9]{40}$`)
+	// for testing
+	readBuildInfo = debug.ReadBuildInfo
+)
 
-// Get tries to fetch the first 7 digitals of GitHub commit ID from HEAD file in
-// KO_DATA_PATH. If it fails, it returns the error it gets.
+// Get returns the 'vcs.revision' property from the embedded build information
+// This function will return an error if value is not a valid Git SHA
+//
+// The result will have a '-dirty' suffix if the workspace was not clean
 func Get() (string, error) {
```

Comment:
Can we have a `changeset.MustGet` that panics if it can't detect the version? I'm updating some callsites to get this once at `init`-time, and they're all just `log.Fatal`ing if this returns an error. Or should this return `Unknown` if it can't read build info, and just never return an error?

Reply:
> Or should this return Unknown if it can't read build info, and just never return an error?

I was thinking about this - let me do it in a follow up